src/library.cpp

```diff
@@ -2587,6 +2605,94 @@ void lammps_scatter_atoms_subset(void *handle, char *name, int type, int count,
   END_CAPTURE
 }
 
+/* ----------------------------------------------------------------------
+   Contributing author: Robert Meissner (TUHH & HZG, Hamburg, germany)
+   gather bonded info for all atoms
+     return it in user-allocated data
+   data will be ordered by atom ID
+     requirement for consecutive atom IDs (1 to N)
+   return atom-based values in 1d data, ordered by atom ID
+     e.g. b[0],b[1],x[0][2],x[1][0],x[1][1],x[1][2],x[2][0],...
+     data must be pre-allocated by caller to correct length
+     correct length = Nbonds, as queried by get_nbonds()
+   method:
+     alloc and zero Nbond length vector
+     loop over Nlocal to fill vector with my values
+     Allreduce to sum vector into data across all procs
+------------------------------------------------------------------------- */
+
+void lammps_gather_bonds(void *handle, void *data)
+{
+  LAMMPS *lmp = (LAMMPS *) handle;
+
+  BEGIN_CAPTURE
+  {
+#if defined(LAMMPS_BIGBIG)
+    lmp->error->all(FLERR,"Library function lammps_gather_atoms() "
+                    "is not compatible with -DLAMMPS_BIGBIG");
+#else
+    int i,j;
+
+    // error if tags are not defined or not consecutive
+    // NOTE: test that name = image or ids is not a 64-bit int in code?
+
+    int flag = 0;
+    if (lmp->atom->tag_enable == 0 || lmp->atom->tag_consecutive() == 0)
+      flag = 1;
+    if (lmp->atom->natoms > MAXSMALLINT) flag = 1;
+    if (flag) {
+      if (lmp->comm->me == 0)
+        lmp->error->warning(FLERR,"Library error in lammps_gather_bonds");
+      return;
+    }
+    
+    int natoms = static_cast<int> (lmp->atom->natoms);
+    int nbonds = static_cast<int> (lmp->atom->nbonds);
+    if (!lmp->force->newton_bond) nbonds *= 2;
+    
+    int *offset;
+    lmp->memory->create(offset,natoms,"lib/gather:copy");
+    for (i = 0; i < natoms; i++) offset[i] = 0;
+    
+    tagint *tag = lmp->atom->tag;
+    int nlocal = lmp->atom->nlocal;
+    int *num_bond = lmp->atom->num_bond;
+  
+    for (i = 0; i < nlocal; i++) offset[tag[i]-1] = num_bond[i];
```

Comment:
The code will segfault here when `lmp->atom->molecular != Atom::MOLECULAR`. There will be no bonds with `Atom::ATOMIC` and there will be no arrays with `Atom::TEMPLATE` as the bond information in that case is provided by molecule templates.

Reply:
Checking now in the beginning of `gather_bonds()` if `Atom::MOLECULAR` is `TRUE` to avoid this issue.